README.md
fabistrano
============

Fabistrano is a set of [Fabric](http://docs.fabfile.org/en/1.6/) tasks that allows you to do [Capistrano](https://github.com/capistrano/capistrano) style deployments for python web apps.

## What is a _capistrano style_ deployment?

Capistrano deploys code under versioned directories and then symlinks the latest one as the *current* release.

A sample structure would look like this:

  ```
    app.com
    |
    | - releases
      | - 1366924759
      | - 1366927898
      | - 1367155641
    | - shared
    | - current
  ```

## Why use this style of deployment?

Using fabistrano you get the following benefits:

- Scripted and ordered deployments
- Very low downtime when deploying new code
- Easy roll backs

## Usage

1. Install fabistrano:

` pip install fabistrano `

2. In your `fabfile.py` you need to import fabistrano and set the environment variables:

  ```
    from fabistrano import deploy

    
    env.hosts = ["HOST"] # Replace with your host name or IP
    env.base_dir = '/www' # Set to your app's directory
    env.app_name = 'app_name.com' # This will deploy the app to /www/app_name.com/
    env.git_clone = 'GIT_PATH' # Your git url

    env.restart_cmd = 'kill -HUP `supervisorctl pid gunicorn`' # Restart command
    # or
    # env.wsgi_path = "app_name/apache.wsgi" # Relative path to the wsgi file to be touched on restart
  ```


3. Run `setup` to create the directory structure:

  ` fab deploy.setup `

4. Deploy the app:

  ` fab deploy `

5. Setup your web server to point to the `current` directory.


And you should be good to go!

## Current status

This tool is under active development and you might see errors.

## License

The template itself is available under the "Simplified" BSD license and can be
freely modified under the terms of that license. You can see the 
[full license text](https://github.com/dlapiduz/fabistrano/blob/master/LICENSE>)
in the Github repository for additional detail.

Applications created using this template are not considered derivatives works.
Applications created using this template can be freely licensed under whatever
you as the author chooses. They may be either open or closed source.



CHANGELOG.md
# Fabistrano changelog

## Unreleased

### Added

- This changelog.
- Ability to avoid using sudo.

### Fixed

- Remove committed `.pyc` files and add `*.pyc` to `.gitignore`.
- Remove committed distribution packages.
- Typos in `README.md`.

### Removed

- Unused `VERSION` variable in `fabistrano.deploy`.


## 0.3 - 2013-04-29

### Changed

- Rename `fabistrano.fabistrano` module to `fabistrano.deploy`.
- Change `fabistrano.helpers.dir_exists` to be a no-op.

### Fixed

- Sources package fixed to properly include source files.


## 0.2 - 2013-04-29 [YANKED]

### Changed

- Rename `src` module to `fabistrano`


## 0.1 - 2013-04-28 [YANKED]

### Added

- Initial Release


fabistrano/deploy.py
from fabric.api import env, sudo, run, put, task
from fabistrano.helpers import with_defaults


env.timeout = 6000

def sudo_run(*args, **kwargs):
    if env.use_sudo:
        sudo(*args, **kwargs)
    else:
        run(*args, **kwargs)

@task
@with_defaults
def restart():
    """Restarts your application"""
    try:
        run("touch %(current_release)s/%(wsgi_path)s" % \
                { 'current_release': env.current_release,
                  'wsgi_path': env.wsgi_path })
    except AttributeError:
        try:
            sudo_run(env.restart_cmd)
        except AttributeError:
            pass

@with_defaults
def permissions():
    """Make the release group-writable"""
    sudo_run("chown -R %(user)s:%(group)s %(domain_path)s" % 
            { 'domain_path':env.domain_path,
              'user': env.remote_owner,
              'group': env.remote_group })
    sudo_run("chmod -R g+w %(domain_path)s" % { 'domain_path':env.domain_path })

@task
@with_defaults
def setup():
    """Prepares one or more servers for deployment"""
    sudo_run("mkdir -p %(domain_path)s/{releases,shared}" % { 'domain_path':env.domain_path })
    sudo_run("mkdir -p %(shared_path)s/{system,log}" % { 'shared_path':env.shared_path })
    permissions()

@with_defaults
def checkout():
    """Checkout code to the remote servers"""
    from time import time
    env.current_release = "%(releases_path)s/%(time).0f" % { 'releases_path':env.releases_path, 'time':time() }
    run("cd %(releases_path)s; git clone -b %(git_branch)s -q %(git_clone)s %(current_release)s" % \
        { 'releases_path':env.releases_path,
          'git_clone':env.git_clone,
          'current_release':env.current_release,
          'git_branch':env.git_branch })

@task
def update():
    """Copies your project and updates environment and symlink"""
    update_code()
    update_env()
    symlink()
    set_current()
    permissions()

@task
def update_code():
    """Copies your project to the remote servers"""
    checkout()
    permissions()

@with_defaults
def symlink():
    """Updates the symlink to the most recently deployed version"""
    run("ln -nfs %(shared_path)s/log %(current_release)s/log" % { 'shared_path':env.shared_path, 'current_release':env.current_release })

@with_defaults
def set_current():
    """Sets the current directory to the new release"""
    run("ln -nfs %(current_release)s %(current_path)s" % { 'current_release':env.current_release, 'current_path':env.current_path })

@with_defaults
def update_env():
    """Update servers environment on the remote servers"""
    sudo_run("cd %(current_release)s; pip install -r requirements.txt" % { 'current_release':env.current_release })
    permissions()

@task 
@with_defaults
def cleanup():
    """Clean up old releases"""
    if len(env.releases) > 3:
        directories = env.releases
        directories.reverse()
        del directories[:3]
        env.directories = ' '.join([ "%(releases_path)s/%(release)s" % { 'releases_path':env.releases_path, 'release':release } for release in directories ])
        run("rm -rf %(directories)s" % { 'directories':env.directories })

@with_defaults
def rollback_code():
    """Rolls back to the previously deployed version"""
    if len(env.releases) >= 2:
        env.current_release = env.releases[-1]
        env.previous_revision = env.releases[-2]
        env.current_release = "%(releases_path)s/%(current_revision)s" % { 'releases_path':env.releases_path, 'current_revision':env.current_revision }
        env.previous_release = "%(releases_path)s/%(previous_revision)s" % { 'releases_path':env.releases_path, 'previous_revision':env.previous_revision }
        run("rm %(current_path)s; ln -s %(previous_release)s %(current_path)s && rm -rf %(current_release)s" % { 'current_release':env.current_release, 'previous_release':env.previous_release, 'current_path':env.current_path })

@task
def rollback():
    """Rolls back to a previous version and restarts"""
    rollback_code()
    restart()

@task(default=True)
def deploy():
    """Deploys your project. This calls both `update' and `restart'"""
    update()
    restart()



fabistrano/helpers.py
import functools
from fabric.api import run, env


def dir_exists(dir):
    # return run('[ -d %s ] && echo 1 || echo 0' % dir) == '1'
    pass

def with_defaults(func):
    """A decorator that sets all defaults for a task."""

    @functools.wraps(func)
    def decorated(*args, **kwargs):
        env.setdefault('use_sudo', True)
        env.setdefault('git_branch', 'master')
        env.setdefault('python_bin', 'python')
        env.setdefault('remote_owner', 'www-data')
        env.setdefault('remote_group', 'www-data')


        env.setdefault('domain_path', "%(base_dir)s/%(app_name)s" % \
                                  { 'base_dir':env.base_dir,
                                    'app_name':env.app_name })
        env.setdefault('current_path', "%(domain_path)s/current" % \
                                       { 'domain_path':env.domain_path })
        env.setdefault('releases_path', "%(domain_path)s/releases" % \
                                        { 'domain_path':env.domain_path })
        env.setdefault('shared_path', "%(domain_path)s/shared" % \
                                      { 'domain_path':env.domain_path })
        if not env.has_key('releases'):
            if dir_exists(env.releases_path):
                env.releases = sorted(run('ls -x %(releases_path)s' % { 'releases_path':env.releases_path }).split())

                if len(env.releases) >= 1:
                    env.current_revision = env.releases[-1]
                    env.current_release = "%(releases_path)s/%(current_revision)s" % \
                                                { 'releases_path':env.releases_path, 
                                                  'current_revision':env.current_revision }
                if len(env.releases) > 1:
                    env.previous_revision = env.releases[-2]
                    env.previous_release = "%(releases_path)s/%(previous_revision)s" % \
                                                { 'releases_path':env.releases_path, 
                                                  'previous_revision':env.previous_revision }
        return func(*args, **kwargs)
    return decorated


src/fabistrano.py
from fabric.api import env, sudo, run, put, task


VERSION = "0.1"

env.timeout = 6000

@task
@with_defaults
def restart():
    """Restarts your application"""
    try:
        run("touch %(current_release)s/%(wsgi_path)s" % \
                { 'current_release': env.current_release,
                  'wsgi_path': env.wsgi_path })
    except AttributeError:
        try:
            sudo(env.restart_cmd)
        except AttributeError:
            pass

@with_defaults
def permissions():
    """Make the release group-writable"""
    sudo("chown -R %(user)s:%(group)s %(domain_path)s" % 
            { 'domain_path':env.domain_path,
              'user': env.remote_owner,
              'group': env.remote_group })
    sudo("chmod -R g+w %(domain_path)s" % { 'domain_path':env.domain_path })

@task
@with_defaults
def setup():
    """Prepares one or more servers for deployment"""
    sudo("mkdir -p %(domain_path)s/{releases,shared}" % { 'domain_path':env.domain_path })
    sudo("mkdir -p %(shared_path)s/{system,log}" % { 'shared_path':env.shared_path })
    permissions()

@with_defaults
def checkout():
    """Checkout code to the remote servers"""
    from time import time
    env.current_release = "%(releases_path)s/%(time).0f" % { 'releases_path':env.releases_path, 'time':time() }
    run("cd %(releases_path)s; git clone -b %(git_branch)s -q %(git_clone)s %(current_release)s" % \
        { 'releases_path':env.releases_path,
          'git_clone':env.git_clone,
          'current_release':env.current_release,
          'git_branch':env.git_branch })

@task
def update():
    """Copies your project and updates environment and symlink"""
    update_code()
    update_env()
    symlink()
    set_current()
    permissions()

@task
def update_code():
    """Copies your project to the remote servers"""
    checkout()
    permissions()

@with_defaults
def symlink():
    """Updates the symlink to the most recently deployed version"""
    run("ln -nfs %(shared_path)s/log %(current_release)s/log" % { 'shared_path':env.shared_path, 'current_release':env.current_release })

@with_defaults
def set_current():
    """Sets the current directory to the new release"""
    run("ln -nfs %(current_release)s %(current_path)s" % { 'current_release':env.current_release, 'current_path':env.current_path })

@with_defaults
def update_env():
    """Update servers environment on the remote servers"""
    sudo("cd %(current_release)s; pip install -r requirements.txt" % { 'current_release':env.current_release })
    permissions()

@task 
@with_defaults
def cleanup():
    """Clean up old releases"""
    if len(env.releases) > 3:
        directories = env.releases
        directories.reverse()
        del directories[:3]
        env.directories = ' '.join([ "%(releases_path)s/%(release)s" % { 'releases_path':env.releases_path, 'release':release } for release in directories ])
        run("rm -rf %(directories)s" % { 'directories':env.directories })

@with_defaults
def rollback_code():
    """Rolls back to the previously deployed version"""
    if len(env.releases) >= 2:
        env.current_release = env.releases[-1]
        env.previous_revision = env.releases[-2]
        env.current_release = "%(releases_path)s/%(current_revision)s" % { 'releases_path':env.releases_path, 'current_revision':env.current_revision }
        env.previous_release = "%(releases_path)s/%(previous_revision)s" % { 'releases_path':env.releases_path, 'previous_revision':env.previous_revision }
        run("rm %(current_path)s; ln -s %(previous_release)s %(current_path)s && rm -rf %(current_release)s" % { 'current_release':env.current_release, 'previous_release':env.previous_release, 'current_path':env.current_path })

@task
def rollback():
    """Rolls back to a previous version and restarts"""
    rollback_code()
    restart()

@task(default=True)
def deploy():
    """Deploys your project. This calls both `update' and `restart'"""
    update()
    restart()

